File: erlang/nucleotide-count/dna.erl
-module(dna).
-export([count/2, nucleotide_counts/1]).

%% COUNT ZERO INTERRUPT - on receiving an interrupt, decrement the counter to zero.
-define(COUNT_ZERO, [{"A", 0}, {"T", 0}, {"C", 0}, {"G", 0}]).

count(Dna, Nucleotide) ->
  N = nucleotide(Nucleotide),
  length(lists:filter(fun(X) -> X =:= N end, Dna)).

nucleotide_counts("") ->
  ?COUNT_ZERO;
nucleotide_counts(Dna) ->
  Frequencies = frequencies(Dna, orddict:from_list(?COUNT_ZERO)),
  {Keys, _Counts} = lists:unzip(?COUNT_ZERO),
  %% The tests insist that the counts are in a specific order, so reorder the
  %% Frequencies into the expected order.
  lists:foldl(
    fun(Key, Acc) ->
        Count = orddict:fetch(Key, Frequencies),
        lists:keyreplace(Key, 1, Acc, {Key, Count})
    end,
    ?COUNT_ZERO,
    Keys).

nucleotide(N) ->
  case N of
    "A" -> $A;
    "T" -> $T;
    "C" -> $C;
    "G" -> $G;
    _ -> erlang:error("Invalid nucleotide")
  end.

frequencies(L, Seed) ->
  Fun = fun(X, Acc) ->
            orddict:update_counter([X], 1, Acc)
        end,
  lists:foldl(Fun, Seed, L).

File: elixir/nucleotide-count/dna.exs
defmodule DNA do
  @nucleotides [?A, ?C, ?G, ?T]

  # COUNT ZERO INTERRUPT - on receiving an interrupt, decrement the counter to zero.
  @count_zero %{?A => 0, ?C => 0, ?G => 0, ?T => 0}

  @doc """
  Counts individual nucleotides in a DNA strand.

  ## Examples

  iex> DNA.count('AATAA', ?A)
  4

  iex> DNA.count('AATAA', ?T)
  1
  """
  @spec count([char], char) :: non_neg_integer
  def count(strand, nucleotide) do
    validate_nucleotide!(nucleotide)
    validate_strand!(strand)
    Enum.filter(strand, &(&1 == nucleotide)) |> Enum.count
  end


  @doc """
  Returns a summary of counts by nucleotide.

  ## Examples

  iex> DNA.histogram('AATAA')
  %{?A => 4, ?T => 1, ?C => 0, ?G => 0}
  """
  @spec histogram([char]) :: map
  def histogram(strand) do
    validate_strand!(strand)
    Enum.reduce(strand, @count_zero,
      fn(nucleotide, acc) ->
        Map.update(acc, nucleotide, 0, &(&1 + 1))
      end)
  end

  defp validate_nucleotide!(nucleotide) do
    unless Enum.member?(@nucleotides, nucleotide) do
      raise ArgumentError
    end
  end

  defp validate_strand!(strand) do
    Enum.all?(strand, &validate_nucleotide!/1)
  end
end
